### PR TITLE
[8.0] account_payment_return

### PR DESCRIPTION
**Pull request by pedrobaeza**
# 
# Customer Payment Orders Returns

This module implements the return of customer receivables and allows to catch
several reconciled account move lines, to give back to a state where the debt
is still open, and letting history of it.

This module can be extended adding importers that fills automatically the
full payment return record.
# Usage

Go to Accounting > Customers > Customer Payment Returns, and create a new
record, registering on each line a paid (reconciled) receivable journal item,
entering the amount that is going to be returned.

Then, press button "Confirm" to create a new move line that removes the
balance from the bank journal and reconcile items together to show payment
history through it.
# Known issues / Roadmap
- Add a button to see the created move.
- Allow to add a commission amount on each line.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Superseeds https://github.com/OCA/bank-payment/pull/125

File: account_payment_return/README.rst [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+   :alt: License: AGPL-3
+
+===============================
+Customer Payment Orders Returns
+===============================
+
+This module implements the return of customer receivables and allows to catch
+several reconciled account move lines, to give back to a state where the debt
+is still open, and letting history of it.
+
```
Comment (elicoidal) [inline account_payment_return/README.rst:12]: Suggestion:
This module implements customer receivables payback and allows to send 
related reconciled account move lines back to a state where 
the debt is still open, and letting history of it.

File: account_payment_return/README.rst [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+This module can be extended adding importers that fills automatically the
```
Comment (elicoidal) [inline account_payment_return/README.rst]: s/fills automatically/automatically fills

File: account_payment_return/README.rst [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+full payment return record.
```
Comment (elicoidal) [inline account_payment_return/README.rst]: not sure about "return" wording.
Payback? reimbursement?
Anglophone speaker wanted ;)

File: account_payment_return/README.rst [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+Usage
+=====
+
+Go to Accounting > Customers > Customer Payment Returns, and create a new
+record, registering on each line a paid (reconciled) receivable journal item,
```
Comment (elicoidal) [inline account_payment_return/README.rst]: s/record, registering on each line a paid (reconciled) receivable journal item,
/record, register on each line a paid (reconciled) receivable journal item,

File: account_payment_return/README.rst [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+entering the amount that is going to be returned.
```
Comment (elicoidal) [inline account_payment_return/README.rst]: s/entering the amount that is going to be returned.
/and input the amount that is going to be returned.

File: account_payment_return/README.rst [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+Then, press button "Confirm" to create a new move line that removes the
```
Comment (elicoidal) [inline account_payment_return/README.rst]: s/Then/Next

File: account_payment_return/models/account_invoice.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+from openerp import models, fields
+
+
+class account_invoice(models.Model):
+    _inherit = "account.invoice"
+
+    payment_returned = fields.Boolean('Payment returned', help='Invoice '
```
Comment (elicoidal) [inline account_payment_return/models/account_invoice.py]: s/Payment returned/Returned Payment

File: account_payment_return/models/payment_return.py [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    company_id = fields.Many2one(
+        'res.company', 'Company', required=True,
+        help="Company",
+        states={'done': [('readonly', True)],
+                'cancelled': [('readonly', True)]},
+        default=lambda self: self.env['res.company']._company_default_get(
+            'account'))
+    date = fields.Date(
+        string='Return date',
+        help="This date will be used as the date for the account entry.",
```
Comment (elicoidal) [inline account_payment_return/models/payment_return.py]: s/This date will be used as the date for the account entry
/This date will be used as accounting entry date

File: account_payment_return/models/payment_return.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+        invoices = self.env['account.invoice'].search(
+            [('move_id', 'in', invoice_moves.ids)])
+        return invoices
+
+    @api.one
+    def action_confirm(self):
+        # Check for incomplete lines
+        if any(not x.move_line_id for x in self.line_ids):
+            raise UserError(
+                _("You must complete all moves references in the payment "
```
Comment (elicoidal) [inline account_payment_return/models/payment_return.py]: s/complete/input

File: account_payment_return/models/payment_return.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+class PaymentReturnLine(models.Model):
+    _name = "payment.return.line"
+    _description = 'Payment return lines'
+
+    return_id = fields.Many2one(
+        comodel_name='payment.return', string='Payment return',
+        required=True, ondelete='cascade')
+    concept = fields.Char(
+        string='Concept',
+        help="Readed from imported file. Only for reference.")
```
Comment (elicoidal) [inline account_payment_return/models/payment_return.py]: s/readed/Read

File: account_payment_return/models/payment_return.py [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    reason = fields.Char(
+        string='Return reason', readonly=True,
+        help="Readed from imported file. Only for reference.")
```
Comment (elicoidal) [inline account_payment_return/models/payment_return.py]: s/Readed/Read

File: account_payment_return/models/payment_return.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    move_line_id = fields.Many2one(
+        comodel_name='account.move.line', string='Payment Reference')
+    date = fields.Date(
+        string='Return date', readonly=True,
+        help="Readed from imported file. Only for reference.",
```
Comment (elicoidal) [inline account_payment_return/models/payment_return.py]: Read

File: account_payment_return/models/payment_return.py [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+        default=lambda x: fields.Date.today())
+    partner_name = fields.Char(
+        string='Partner name', readonly=True,
+        help="Readed from imported file. Only for reference.")
```
Comment (elicoidal) [inline account_payment_return/models/payment_return.py]: Read

**Comment by pedrobaeza** [comment]
@elicoidal, thanks for you review. I have changed all the referred terms. About the main term, I think "returned payment" seems a correct one searching for occurrences in a searcher.

**Comment by elicoidal** [comment]
@pedrobaeza Returned payment sounds better.
I would suggest to change the name of the module accordingly to account_returned_payment

**Comment by pedrobaeza** [comment]
But the module is named accordingly: for having a returned payment, you have to make a payment return.

**Comment by elicoidal** [comment]
@pedrobaeza Valid.

**Comment by oihane** [comment]
:+1:

File: account_payment_return/__openerp__.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+              'Odoo Community Association (OCA)',
+    'website': 'http://www.serviciosbaeza.com',
+    'data': [
+        'security/ir.model.access.csv',
+        'security/account_payment_return_security.xml',
+        'views/payment_return_view.xml',
+        'data/ir_sequence_data.xml',
+        'views/account_invoice_view.xml',
+    ],
+    'demo': [],
```
Comment (oihane) [inline account_payment_return/__openerp__.py]: You don't need this key

**Comment by pedrobaeza** [comment]
@sergio-incaser, can you review?

**Comment by carlosdauden** [comment]
:+1: 
We have little chages as bank charge or change payment type of "returned" invoice. ¿Whould be do a PR over @pedrobaeza branch or we wait to PR after merge this?

**Comment by pedrobaeza** [comment]
Please make a PR over my branch

**Comment by carlosdauden** [comment]
Travis fails unrelated. (account_payment_order_to_voucher tests)

**Comment by pedrobaeza** [comment]
OK, then it's time to merge this one and continue with other improvements over it. Thanks to all reviewers and contributors.

**Pull request merged**